<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Kristy Ou">

<title>Numbat haplotype-aware CNV analysis at single cell resolution</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="XX2_numbat.mutation_files/libs/clipboard/clipboard.min.js"></script>
<script src="XX2_numbat.mutation_files/libs/quarto-html/quarto.js"></script>
<script src="XX2_numbat.mutation_files/libs/quarto-html/popper.min.js"></script>
<script src="XX2_numbat.mutation_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="XX2_numbat.mutation_files/libs/quarto-html/anchor.min.js"></script>
<link href="XX2_numbat.mutation_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="XX2_numbat.mutation_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="XX2_numbat.mutation_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="XX2_numbat.mutation_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="XX2_numbat.mutation_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#running-numbat" id="toc-running-numbat" class="nav-link active" data-scroll-target="#running-numbat">Running Numbat</a>
  <ul class="collapse">
  <li><a href="#notes-on-preparing-data" id="toc-notes-on-preparing-data" class="nav-link" data-scroll-target="#notes-on-preparing-data">Notes on preparing data</a></li>
  <li><a href="#important-parameters" id="toc-important-parameters" class="nav-link" data-scroll-target="#important-parameters">Important parameters</a></li>
  <li><a href="#benchmark-1" id="toc-benchmark-1" class="nav-link" data-scroll-target="#benchmark-1">Benchmark 1</a></li>
  <li><a href="#benchmark-2" id="toc-benchmark-2" class="nav-link" data-scroll-target="#benchmark-2">Benchmark 2</a></li>
  <li><a href="#benchmark-3" id="toc-benchmark-3" class="nav-link" data-scroll-target="#benchmark-3">Benchmark 3</a></li>
  </ul></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results">Results</a>
  <ul class="collapse">
  <li><a href="#compare-pseudobulked-clones-output-of-run_numbat" id="toc-compare-pseudobulked-clones-output-of-run_numbat" class="nav-link" data-scroll-target="#compare-pseudobulked-clones-output-of-run_numbat">Compare pseudobulked clones (output of <code>run_numbat</code>)</a></li>
  <li><a href="#single-cell-posterior-probabilities-for-each-cnv" id="toc-single-cell-posterior-probabilities-for-each-cnv" class="nav-link" data-scroll-target="#single-cell-posterior-probabilities-for-each-cnv">Single cell posterior probabilities for each CNV</a></li>
  <li><a href="#genotypes-of-clones" id="toc-genotypes-of-clones" class="nav-link" data-scroll-target="#genotypes-of-clones">Genotypes of clones</a></li>
  </ul></li>
  <li><a href="#annotate-cells" id="toc-annotate-cells" class="nav-link" data-scroll-target="#annotate-cells">Annotate cells</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Numbat haplotype-aware CNV analysis at single cell resolution</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Kristy Ou </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p><strong>Packages</strong></p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Signac)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse, <span class="at">quietly =</span> T)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(shiny)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(numbat)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(furrr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(future.apply)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"_theme.R"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># always set seed!!!</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">105</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(<span class="fu">list</span>(</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tweak</span>(multisession, <span class="at">workers =</span> <span class="fu">availableCores</span>() <span class="sc">%/%</span> <span class="dv">4</span>),</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  sequential</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><strong>Data import</strong></p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>LN_ID <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LN0025"</span>,<span class="st">"LN0177"</span>,<span class="st">"LN0193"</span>,<span class="st">"LN0438"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>wnn <span class="ot">=</span> qs<span class="sc">::</span><span class="fu">qread</span>(<span class="st">"rdata/02.04_wnn.qs"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">=</span> wnn<span class="sc">@</span>meta.data</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># embeddings for plotting</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>embeds <span class="ot">=</span> <span class="fu">cbind</span>(</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(wnn,<span class="st">"atacUMAP"</span>),</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(wnn,<span class="st">"sctUMAP"</span>),</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(wnn,<span class="st">"wnnUMAP"</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># save together? qsavem?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="running-numbat" class="level2">
<h2 class="anchored" data-anchor-id="running-numbat">Running Numbat</h2>
<section id="notes-on-preparing-data" class="level3">
<h3 class="anchored" data-anchor-id="notes-on-preparing-data">Notes on preparing data</h3>
<ul>
<li>Counted alleles and phased SNPs in bam files using <code>scripts/03_prelim.annotation/XX_prepare.numbat.sh</code></li>
<li>After SNP phasing, prepared the following input for <code>run_numbat()</code>:
<ol type="1">
<li>output of allele counts</li>
<li>count matrix (gene x cell)</li>
<li>reference expression matrix</li>
</ol></li>
</ul>
</section>
<section id="important-parameters" class="level3">
<h3 class="anchored" data-anchor-id="important-parameters">Important parameters</h3>
<ul>
<li>Other notes for <code>run_numbat()</code> , see “Run Parameters”: <a href="https://kharchenkolab.github.io/numbat/articles/numbat.html" class="uri">https://kharchenkolab.github.io/numbat/articles/numbat.html</a></li>
<li>Lower <code>t</code> (transition property) is better for resolving additional subclonal events</li>
<li>Lower <code>max_cost</code> generates more clones. Default: <code>ncol(count_mat) * tau</code></li>
<li>Default <code>tau</code> = 0.3, set lower to find more clones</li>
<li>See <code>scripts/03_prelim.annotation/XX1_run.numbat.slurmR</code></li>
</ul>
</section>
<section id="benchmark-1" class="level3">
<h3 class="anchored" data-anchor-id="benchmark-1">Benchmark 1</h3>
<ul>
<li>Allele / Count input: all cells (except singleton, mitotic), separated by patient</li>
<li>Expression reference: all cells minus singleton, mitotic, BCL2.hi B cells; collective</li>
<li><code>t</code> transition probabiltiy: 1e-6</li>
<li><code>tau</code> = 0.25</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define cells to keep ----</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">=</span> <span class="fu">Cells</span>(wnn)[<span class="sc">!</span>(<span class="fu">Idents</span>(wnn) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"singleton"</span>,<span class="st">"mitotic"</span>))]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># get allele counts, was determined in all cells ----</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>dir<span class="ot">=</span><span class="st">"/scratch/kou/composite/numbat/"</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>files <span class="ot">=</span> <span class="fu">list.files</span>(<span class="fu">paste0</span>(dir,LN_ID), <span class="at">pattern =</span> <span class="st">"tsv.gz$"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>alleleCounts <span class="ot">=</span> <span class="fu">lapply</span>(files, read_tsv)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="do">## modify cell id to match seurat, filter for "keep" cells</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>alleleCounts <span class="ot">=</span> <span class="fu">map2</span>(</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  LN_ID,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  alleleCounts, </span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> .y <span class="sc">%&gt;%</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">cell =</span> <span class="fu">paste0</span>(.x,<span class="st">"_"</span>,cell)) <span class="sc">%&gt;%</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(cell <span class="sc">%in%</span> keep)</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co"># get count matrix ----</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="do">## count matrix, split by sample</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>countMat <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="do">## remove mitotic and singletons</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>countMat <span class="ot">=</span> countMat[,<span class="fu">colnames</span>(countMat) <span class="sc">%in%</span> keep]</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="do">## split count matrix by sample</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>countMat <span class="ot">=</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(x){</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> countMat[,<span class="fu">grepl</span>(LN_ID[[x]],<span class="fu">colnames</span>(countMat))]</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># round mat to integer, otherwise error later</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> <span class="fu">round</span>(mat)</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mat)</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a><span class="co"># make collective reference matrix ----</span></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a><span class="do">## remove bcl2.hi (likely) tumor cells</span></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>noBCL2hi <span class="ot">=</span> <span class="fu">levels</span>(wnn)[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"BCL2.hi|singleton|mitotic"</span>,<span class="fu">levels</span>(wnn))]</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a><span class="do">## get counts, subset cells </span></span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>refCounts <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[, <span class="fu">Idents</span>(wnn) <span class="sc">%in%</span> noBCL2hi]</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a><span class="do">## create anno</span></span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>refAnno <span class="ot">=</span> meta <span class="sc">%&gt;%</span> </span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cell =</span> <span class="fu">paste0</span>(orig.ident,<span class="st">"_"</span>,cell_ids)) <span class="sc">%&gt;%</span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cell <span class="sc">%in%</span> <span class="fu">colnames</span>(refCounts)) <span class="sc">%&gt;%</span></span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(cell, <span class="at">group =</span> prelim_anno)</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a><span class="do">## aggregate counts</span></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a>ref_internal <span class="ot">=</span> <span class="fu">aggregate_counts</span>(</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>  <span class="at">count_mat =</span> refCounts, </span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>  <span class="at">annot =</span> refAnno, </span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a>  <span class="at">normalized =</span> T, </span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> F</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a><span class="co"># run numbat ----</span></span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a><span class="do">## should run on slurm next time</span></span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a>outdir<span class="ot">=</span><span class="st">"rdata/numbat/benchmark1/"</span></span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a>nbResults <span class="ot">=</span> <span class="fu">map</span>(</span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(countMat),</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">run_numbat</span>(</span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">count_mat =</span> countMat[[.x]],</span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">lambdas_ref =</span> ref_internal,</span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">df_allele =</span> alleleCounts[[.x]],</span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">genome =</span> <span class="st">"hg38"</span>,</span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a>    <span class="at">out_dir =</span> <span class="fu">paste0</span>(outdir,LN_ID[[.x]]),</span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a>    <span class="at">t =</span> <span class="fl">1e-06</span>, </span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a>    <span class="at">gamma =</span> <span class="dv">20</span>,</span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_cells =</span> <span class="dv">30</span>, </span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_depth =</span> <span class="dv">0</span>,</span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a>    <span class="at">multi_allelic =</span> T,</span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a>    <span class="at">tau =</span> <span class="fl">0.25</span>,</span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>,</span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores_nni =</span> <span class="dv">1</span></span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="benchmark-2" class="level3">
<h3 class="anchored" data-anchor-id="benchmark-2">Benchmark 2</h3>
<ul>
<li>Allele / Count input: all cells (including singleton and mitotic!), separated by patient</li>
<li>Expression reference: all cells minus singleton, mitotic, BCL2.hi B cells; used patient specific reference</li>
<li><code>t</code> transition probabiltiy: 1e-6</li>
<li><code>tau</code> = 0.25</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- get allele counts, was determined in all cells ----</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>dir<span class="ot">=</span><span class="st">"/scratch/kou/composite/numbat/"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>files <span class="ot">=</span> <span class="fu">list.files</span>(<span class="fu">paste0</span>(dir,LN_ID), <span class="at">pattern =</span> <span class="st">"tsv.gz$"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>alleleCounts2 <span class="ot">=</span> <span class="fu">lapply</span>(files, read_tsv)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="do">## modify cell id to match seurat</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>alleleCounts2 <span class="ot">=</span> <span class="fu">map2</span>(</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  LN_ID,</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  alleleCounts2, </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> .y <span class="sc">%&gt;%</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">cell =</span> <span class="fu">paste0</span>(.x,<span class="st">"_"</span>,cell))</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- get count matrix ----</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="do">## count matrix, split by sample</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>countMat2 <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="do">## split count matrix by sample</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>countMat2 <span class="ot">=</span> <span class="fu">future_lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(x){</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> countMat2[,<span class="fu">grepl</span>(LN_ID[[x]],<span class="fu">colnames</span>(countMat2))]</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># round mat to integer, otherwise error later</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> <span class="fu">round</span>(mat)</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mat)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- make patient specific reference matrix ----</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="do">## remove bcl2.hi and mitotic bc (likely) tumor cells</span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>noBCL2hi <span class="ot">=</span> <span class="fu">levels</span>(wnn)[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"BCL2.hi|mitotic"</span>,<span class="fu">levels</span>(wnn))]</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a><span class="do">## get counts, subset cells </span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>refCounts2 <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[, <span class="fu">Idents</span>(wnn) <span class="sc">%in%</span> noBCL2hi]</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="do">## split ref matrix by sample</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>refCounts2 <span class="ot">=</span> <span class="fu">future_lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(x){</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> refCounts2[,<span class="fu">grepl</span>(LN_ID[[x]],<span class="fu">colnames</span>(refCounts2))]</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mat)</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="do">## create anno </span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>refAnno2 <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(prelim_anno <span class="sc">%in%</span> noBCL2hi) <span class="sc">%&gt;%</span></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"cell"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(orig.ident) <span class="sc">%&gt;%</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(orig.ident, cell, <span class="at">group =</span> prelim_anno) <span class="sc">%&gt;%</span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_split</span>(<span class="at">.keep =</span> F)</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a><span class="do">## aggregate counts</span></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>ref2 <span class="ot">=</span> <span class="fu">future_map2</span>(</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>  refCounts2, </span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>  refAnno2,</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">aggregate_counts</span>(</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>    <span class="at">count_mat =</span> .x,</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot =</span> .y, </span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>    <span class="at">normalized =</span> T, </span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> F</span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- run numbat ----</span></span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>outdir<span class="ot">=</span><span class="st">"rdata/numbat/benchmark2/"</span></span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>nbResults2 <span class="ot">=</span> <span class="fu">future_map</span>(</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(countMat2),</span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">run_numbat</span>(</span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>    <span class="at">count_mat =</span> countMat2[[.x]],</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>    <span class="at">lambdas_ref =</span> ref2[[.x]],</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>    <span class="at">df_allele =</span> alleleCounts2[[.x]],</span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>    <span class="at">genome =</span> <span class="st">"hg38"</span>,</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>    <span class="at">out_dir =</span> <span class="fu">paste0</span>(outdir,LN_ID[[.x]]),</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>    <span class="at">t =</span> <span class="fl">1e-06</span>, </span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>    <span class="at">gamma =</span> <span class="dv">20</span>,</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_cells =</span> <span class="dv">30</span>, </span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_depth =</span> <span class="dv">0</span>,</span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>    <span class="at">multi_allelic =</span> T,</span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>    <span class="at">tau =</span> <span class="fl">0.25</span>,</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">4</span>,</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores_nni =</span> <span class="dv">4</span></span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="benchmark-3" class="level3">
<h3 class="anchored" data-anchor-id="benchmark-3">Benchmark 3</h3>
<ul>
<li>reference from rLN? needs to be same anno?</li>
</ul>
</section>
</section>
<section id="results" class="level2">
<h2 class="anchored" data-anchor-id="results">Results</h2>
<p><strong>Import numbat results</strong></p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># initialize numbat class</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>dir1 <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"rdata/numbat/benchmark1/"</span>,LN_ID)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>nbResults1 <span class="ot">=</span> <span class="fu">lapply</span>(dir1, Numbat<span class="sc">$</span>new)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(nbResults1) <span class="ot">=</span> LN_ID</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>dir2 <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"rdata/numbat/benchmark2/"</span>,LN_ID)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>nbResults2 <span class="ot">=</span> <span class="fu">lapply</span>(dir2, Numbat<span class="sc">$</span>new)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(nbResults2) <span class="ot">=</span> LN_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="compare-pseudobulked-clones-output-of-run_numbat" class="level3">
<h3 class="anchored" data-anchor-id="compare-pseudobulked-clones-output-of-run_numbat">Compare pseudobulked clones (output of <code>run_numbat</code>)</h3>
<div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># need absolute paths!</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>LN_ID <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LN0025"</span>,<span class="st">"LN0177"</span>,<span class="st">"LN0193"</span>,<span class="st">"LN0438"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>plotpaths <span class="ot">=</span> purrr<span class="sc">::</span><span class="fu">map</span>(</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  LN_ID, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">file.path</span>(</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    here<span class="sc">::</span><span class="fu">here</span>(),</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"rdata/numbat/benchmark"</span>, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)),</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    .x,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bulk_clones_final.png"</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co"># lapply didn't work here</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">1</span>]])</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">2</span>]])</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">3</span>]])</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">4</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0025, bench. 1</figcaption>
<p><img src="../../rdata/numbat/benchmark1/LN0025/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0025, bench. 2</figcaption>
<p><img src="../../rdata/numbat/benchmark2/LN0025/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0177, bench. 1</figcaption>
<p><img src="../../rdata/numbat/benchmark1/LN0177/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0177, bench. 2</figcaption>
<p><img src="../../rdata/numbat/benchmark2/LN0177/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0193, bench. 1</figcaption>
<p><img src="../../rdata/numbat/benchmark1/LN0193/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0193, bench. 2</figcaption>
<p><img src="../../rdata/numbat/benchmark2/LN0193/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0438, bench. 1</figcaption>
<p><img src="../../rdata/numbat/benchmark1/LN0438/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">LN0438, bench. 2</figcaption>
<p><img src="../../rdata/numbat/benchmark2/LN0438/bulk_clones_final.png" class="img-fluid figure-img" width="1625"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<ul>
<li>With the exception of LN0025, there were no major differences between benchmark 1 and 2.</li>
<li>Continued the rest of the analysis with benchmark 2 because more cells were included in analysis (ie - mitotic and singletons).</li>
</ul>
</section>
<section id="single-cell-posterior-probabilities-for-each-cnv" class="level3">
<h3 class="anchored" data-anchor-id="single-cell-posterior-probabilities-for-each-cnv">Single cell posterior probabilities for each CNV</h3>
<p>[[TO DO: add more detail + plots, or remove]]</p>
<ul>
<li>p_cnv: etc</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get current joint_post</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>jointPost <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  nbResults2, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb) {</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get probabilities for each cnv</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> <span class="fu">as_tibble</span>(nb<span class="sc">$</span>joint_post) <span class="sc">%&gt;%</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">cnv =</span> <span class="fu">factor</span>(<span class="fu">paste0</span>(seg,<span class="st">"_"</span>,cnv_state))) <span class="sc">%&gt;%</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>      <span class="co"># select(</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   cell,cnv,cnv_state,LLR,</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   CHROM,seg_start,seg_end,</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   n_genes,n_snp,</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   p_cnv,p_cnv_x,p_cnv_y</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>      <span class="co"># )</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="co"># add cell annotation</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>jointPost <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(jointPost), </span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) {</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get cell anno</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>    cellanno <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">cell =</span> <span class="fu">rownames</span>(meta),</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">major_type =</span> meta<span class="sc">$</span>major_type,</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>      <span class="at">prelim_anno =</span> meta<span class="sc">$</span>prelim_anno,</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>      <span class="at">tricycle =</span> meta<span class="sc">$</span>tricycle</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> <span class="fu">left_join</span>(jointPost[[x]], cellanno, <span class="at">by =</span> <span class="st">"cell"</span>)</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a><span class="co"># test plot, B cells only</span></span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>jointPost[[<span class="dv">3</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>    <span class="co">#major_type %in% c("T"),</span></span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>    cnv_state <span class="sc">!=</span> <span class="st">"loh"</span>,</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>    LLR <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">bcl2 =</span> <span class="fu">if_else</span>(<span class="fu">grepl</span>(<span class="st">"BCL2.hi"</span>, prelim_anno),<span class="st">"hi"</span>,<span class="st">"lo"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(p_neu)) <span class="sc">+</span></span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>  <span class="co">#geom_density() +</span></span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">color =</span> major_type)) <span class="sc">+</span></span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(cnv), <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="genotypes-of-clones" class="level3">
<h3 class="anchored" data-anchor-id="genotypes-of-clones">Genotypes of clones</h3>
<p>TO DO:</p>
<ul>
<li>change levels to include type of mutation!!</li>
<li>re order levels, change color code!! resize ggplot</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default results ----</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="do">## get clone info</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>clonePost <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  nbResults2,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb) {</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> nb<span class="sc">$</span>clone_post</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>clones <span class="ot">=</span> <span class="fu">do.call</span>(rbind,clonePost)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="do">## checked order first all(clones$cell == rownames(embeds))</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>clones <span class="ot">=</span> <span class="fu">cbind</span>(clones, embeds)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>clones <span class="ot">=</span> <span class="fu">cbind</span>(clones, meta)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="do">## relevel</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>clones<span class="sc">$</span>GT_opt <span class="ot">=</span> <span class="fu">factor</span>(</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  clones<span class="sc">$</span>GT_opt, </span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">""</span>,<span class="st">""</span>)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="do">## genotype color code</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>gtColor <span class="ot">=</span> <span class="fu">hcl.colors</span>(<span class="at">n =</span> <span class="dv">4</span>, <span class="st">"Spectral"</span>)</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="do">## generate random order for plotting</span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>cloneplot1 <span class="ot">=</span> clones[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(clones), <span class="fu">nrow</span>(clones), <span class="at">replace =</span> F),] <span class="sc">%&gt;%</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(wnnUMAP_1, wnnUMAP_2))<span class="sc">+</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">color =</span> GT_opt), </span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.05</span>, </span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>  )<span class="sc">+</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_umap</span>() <span class="sc">+</span></span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#F5F5F5"</span>,gtColor))<span class="sc">+</span></span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"Genotype"</span>) <span class="sc">+</span></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(orig.ident))</span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">"figures/"</span>,<span class="fu">format</span>(<span class="fu">Sys.time</span>(),<span class="st">"%y%m%d"</span>),<span class="st">"_NumbatClonesK1.png"</span>),</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>  cloneplot1, <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">3</span></span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a><span class="co"># recut tree (k = 2) to see if subclones resolve ----</span></span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a><span class="do">## this will change the numbat benchmark2 dir</span></span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>nb_k2 <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>  nbResults2,</span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb){</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get 3 groups from 2 cuts</span></span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>    nb<span class="sc">$</span><span class="fu">cutree</span>(<span class="at">n_cut =</span> <span class="dv">2</span>)</span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(nb)</span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a><span class="do">## get clone info</span></span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a>clonePost_k2 <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a>  nb_k2,</span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb) {</span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> nb<span class="sc">$</span>clone_post</span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a>clones_k2 <span class="ot">=</span> <span class="fu">do.call</span>(rbind,clonePost_k2)</span>
<span id="cb8-66"><a href="#cb8-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-67"><a href="#cb8-67" aria-hidden="true" tabindex="-1"></a><span class="do">## checked order first all(clones$cell == rownames(embeds))</span></span>
<span id="cb8-68"><a href="#cb8-68" aria-hidden="true" tabindex="-1"></a>clones_k2 <span class="ot">=</span> <span class="fu">cbind</span>(clones_k2, embeds)</span>
<span id="cb8-69"><a href="#cb8-69" aria-hidden="true" tabindex="-1"></a>clones_k2 <span class="ot">=</span> <span class="fu">cbind</span>(clones_k2,meta)</span>
<span id="cb8-70"><a href="#cb8-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-71"><a href="#cb8-71" aria-hidden="true" tabindex="-1"></a><span class="do">## genotype color code</span></span>
<span id="cb8-72"><a href="#cb8-72" aria-hidden="true" tabindex="-1"></a>gtColor <span class="ot">=</span> <span class="fu">hcl.colors</span>(<span class="at">n =</span> <span class="dv">8</span>, <span class="st">"Zissou"</span>)</span>
<span id="cb8-73"><a href="#cb8-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-74"><a href="#cb8-74" aria-hidden="true" tabindex="-1"></a><span class="do">## random order for plotting</span></span>
<span id="cb8-75"><a href="#cb8-75" aria-hidden="true" tabindex="-1"></a>cloneplot2 <span class="ot">=</span> clones_k2[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(clones_k2), <span class="fu">nrow</span>(clones_k2), <span class="at">replace =</span> F),] <span class="sc">%&gt;%</span></span>
<span id="cb8-76"><a href="#cb8-76" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(wnnUMAP_1, wnnUMAP_2))<span class="sc">+</span></span>
<span id="cb8-77"><a href="#cb8-77" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(</span>
<span id="cb8-78"><a href="#cb8-78" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">color =</span> GT_opt), </span>
<span id="cb8-79"><a href="#cb8-79" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.05</span>, </span>
<span id="cb8-80"><a href="#cb8-80" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb8-81"><a href="#cb8-81" aria-hidden="true" tabindex="-1"></a>  )<span class="sc">+</span></span>
<span id="cb8-82"><a href="#cb8-82" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_umap</span>() <span class="sc">+</span></span>
<span id="cb8-83"><a href="#cb8-83" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#F5F5F5"</span>,gtColor))<span class="sc">+</span></span>
<span id="cb8-84"><a href="#cb8-84" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"Genotype"</span>) <span class="sc">+</span></span>
<span id="cb8-85"><a href="#cb8-85" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(orig.ident))</span>
<span id="cb8-86"><a href="#cb8-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-87"><a href="#cb8-87" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(</span>
<span id="cb8-88"><a href="#cb8-88" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">"figures/"</span>,<span class="fu">format</span>(<span class="fu">Sys.time</span>(),<span class="st">"%y%m%d"</span>),<span class="st">"_NumbatClonesK2.png"</span>),</span>
<span id="cb8-89"><a href="#cb8-89" aria-hidden="true" tabindex="-1"></a>  cloneplot2, <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">3</span></span>
<span id="cb8-90"><a href="#cb8-90" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">=</span> <span class="fu">list.files</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> here<span class="sc">::</span><span class="fu">here</span>(),</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">pattern =</span> <span class="st">"NumbatClones"</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">recursive =</span> T,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.names =</span> T</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(path[[<span class="dv">1</span>]])</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(path[[<span class="dv">2</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">Genotypes, ncut k: 1</figcaption>
<p><img src="../../figures/231208_NumbatClonesK1.png" class="img-fluid figure-img" width="900"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figcaption class="figure-caption">Genotypes, ncut k: 2</figcaption>
<p><img src="../../figures/231208_NumbatClonesK2.png" class="img-fluid figure-img" width="900"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<ul>
<li>More background in T cells when forcing k = 2 (except possibly for LN0177)</li>
</ul>
<p><strong>Tumor vs non-tumor calls</strong></p>
<p>TO DO:</p>
<ul>
<li>update with updated GT with type of CNV</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># checked same order w/ all()</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>tumorCalls <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">compartment_k1 =</span> clones<span class="sc">$</span>compartment_opt,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">gt_k1 =</span> clones<span class="sc">$</span>GT_opt,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">clone_k1 =</span> clones<span class="sc">$</span>clone_opt,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">compartment_k2 =</span> clones_k2<span class="sc">$</span>compartment_opt,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">gt_k2 =</span> clones_k2<span class="sc">$</span>GT_opt,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">clone_k2 =</span> clones_k2<span class="sc">$</span>clone_opt,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">major_type =</span> meta<span class="sc">$</span>major_type,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">prelim_anno =</span> meta<span class="sc">$</span>prelim_anno,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">orig.ident =</span> meta<span class="sc">$</span>orig.ident</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="co"># pivot</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>tumorCalls <span class="ot">=</span> <span class="fu">pivot_longer</span>(</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>  tumorCalls,</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">cols =</span> <span class="fu">names</span>(tumorCalls)[<span class="fu">grep</span>(<span class="st">"compartment"</span>,<span class="fu">names</span>(tumorCalls))],</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to =</span> <span class="st">"k"</span>,</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to =</span> <span class="st">"call"</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate proportion</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>propTumor <span class="ot">=</span> tumorCalls <span class="sc">%&gt;%</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(orig.ident, k, major_type, call) <span class="sc">%&gt;%</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>(call) <span class="sc">%&gt;%</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prop =</span> n <span class="sc">/</span> <span class="fu">sum</span>(n))</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(propTumor, <span class="fu">aes</span>(k, prop))<span class="sc">+</span></span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">fill =</span> call)) <span class="sc">+</span></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">rows =</span> <span class="fu">vars</span>(orig.ident),</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>             <span class="at">cols =</span> <span class="fu">vars</span>(major_type)) <span class="sc">+</span></span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>TO DO: more systematic way of taking all benchmark, compare n cut 1 vs 2, then assess clones</p>
</section>
</section>
<section id="annotate-cells" class="level2">
<h2 class="anchored" data-anchor-id="annotate-cells">Annotate cells</h2>
<ul>
<li>For now, annotate cells</li>
</ul>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session Info</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.2 (2022-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Rocky Linux 8.8 (Green Obsidian)

Matrix products: default
BLAS/LAPACK: /g/easybuild/x86_64/Rocky/8/haswell/software/FlexiBLAS/3.2.1-GCC-12.2.0/lib64/libflexiblas.so.3.2

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] here_1.0.1        png_0.1-8         rprojroot_2.0.4   digest_0.6.33    
 [5] lifecycle_1.0.4   jsonlite_1.8.8    magrittr_2.0.3    evaluate_0.23    
 [9] rlang_1.1.2       cli_3.6.2         rstudioapi_0.14   vctrs_0.6.5      
[13] rmarkdown_2.25    tools_4.2.2       htmlwidgets_1.6.4 purrr_1.0.2      
[17] xfun_0.41         yaml_2.3.7        fastmap_1.1.1     compiler_4.2.2   
[21] htmltools_0.5.7   knitr_1.45       </code></pre>
</div>
</div>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<ul>
<li>original publication: <a href="https://www.nature.com/articles/s41587-022-01468-y" class="uri">https://www.nature.com/articles/s41587-022-01468-y</a></li>
<li><a href="https://kharchenkolab.github.io/numbat/articles/numbat.html" class="uri">https://kharchenkolab.github.io/numbat/articles/numbat.html</a></li>
</ul>
<!-- -->

</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb13" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Numbat haplotype-aware CNV analysis at single cell resolution"</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>**Packages**</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: setup</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Signac)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse, <span class="at">quietly =</span> T)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(shiny)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(numbat)</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(furrr)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(future.apply)</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"_theme.R"</span>)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a><span class="co"># always set seed!!!</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">105</span>)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(<span class="fu">list</span>(</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tweak</span>(multisession, <span class="at">workers =</span> <span class="fu">availableCores</span>() <span class="sc">%/%</span> <span class="dv">4</span>),</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>  sequential</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>**Data import**</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: data</span></span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a>LN_ID <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LN0025"</span>,<span class="st">"LN0177"</span>,<span class="st">"LN0193"</span>,<span class="st">"LN0438"</span>)</span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a>wnn <span class="ot">=</span> qs<span class="sc">::</span><span class="fu">qread</span>(<span class="st">"rdata/02.04_wnn.qs"</span>)</span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">=</span> wnn<span class="sc">@</span>meta.data</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a><span class="co"># embeddings for plotting</span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a>embeds <span class="ot">=</span> <span class="fu">cbind</span>(</span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(wnn,<span class="st">"atacUMAP"</span>),</span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(wnn,<span class="st">"sctUMAP"</span>),</span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(wnn,<span class="st">"wnnUMAP"</span>)</span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true" tabindex="-1"></a><span class="co"># save together? qsavem?</span></span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-50"><a href="#cb13-50" aria-hidden="true" tabindex="-1"></a><span class="fu">## Running Numbat</span></span>
<span id="cb13-51"><a href="#cb13-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-52"><a href="#cb13-52" aria-hidden="true" tabindex="-1"></a><span class="fu">### Notes on preparing data</span></span>
<span id="cb13-53"><a href="#cb13-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-54"><a href="#cb13-54" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Counted alleles and phased SNPs in bam files using <span class="in">`scripts/03_prelim.annotation/XX_prepare.numbat.sh`</span></span>
<span id="cb13-55"><a href="#cb13-55" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>After SNP phasing, prepared the following input for <span class="in">`run_numbat()`</span>:</span>
<span id="cb13-56"><a href="#cb13-56" aria-hidden="true" tabindex="-1"></a><span class="ss">    1.  </span>output of allele counts</span>
<span id="cb13-57"><a href="#cb13-57" aria-hidden="true" tabindex="-1"></a><span class="ss">    2.  </span>count matrix (gene x cell)</span>
<span id="cb13-58"><a href="#cb13-58" aria-hidden="true" tabindex="-1"></a><span class="ss">    3.  </span>reference expression matrix</span>
<span id="cb13-59"><a href="#cb13-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-60"><a href="#cb13-60" aria-hidden="true" tabindex="-1"></a><span class="fu">### Important parameters</span></span>
<span id="cb13-61"><a href="#cb13-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-62"><a href="#cb13-62" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Other notes for <span class="in">`run_numbat()`</span> , see "Run Parameters": <span class="ot">&lt;https://kharchenkolab.github.io/numbat/articles/numbat.html&gt;</span></span>
<span id="cb13-63"><a href="#cb13-63" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Lower <span class="in">`t`</span> (transition property) is better for resolving additional subclonal events</span>
<span id="cb13-64"><a href="#cb13-64" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Lower <span class="in">`max_cost`</span> generates more clones. Default: <span class="in">`ncol(count_mat) * tau`</span></span>
<span id="cb13-65"><a href="#cb13-65" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Default <span class="in">`tau`</span> = 0.3, set lower to find more clones</span>
<span id="cb13-66"><a href="#cb13-66" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>See <span class="in">`scripts/03_prelim.annotation/XX1_run.numbat.slurmR`</span></span>
<span id="cb13-67"><a href="#cb13-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-68"><a href="#cb13-68" aria-hidden="true" tabindex="-1"></a><span class="fu">### Benchmark 1</span></span>
<span id="cb13-69"><a href="#cb13-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-70"><a href="#cb13-70" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Allele / Count input: all cells (except singleton, mitotic), separated by patient</span>
<span id="cb13-71"><a href="#cb13-71" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Expression reference: all cells minus singleton, mitotic, BCL2.hi B cells; collective</span>
<span id="cb13-72"><a href="#cb13-72" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`t`</span> transition probabiltiy: 1e-6</span>
<span id="cb13-73"><a href="#cb13-73" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`tau`</span> = 0.25</span>
<span id="cb13-74"><a href="#cb13-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-77"><a href="#cb13-77" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-78"><a href="#cb13-78" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: benchmark1_code</span></span>
<span id="cb13-79"><a href="#cb13-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-80"><a href="#cb13-80" aria-hidden="true" tabindex="-1"></a><span class="co"># define cells to keep ----</span></span>
<span id="cb13-81"><a href="#cb13-81" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">=</span> <span class="fu">Cells</span>(wnn)[<span class="sc">!</span>(<span class="fu">Idents</span>(wnn) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"singleton"</span>,<span class="st">"mitotic"</span>))]</span>
<span id="cb13-82"><a href="#cb13-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-83"><a href="#cb13-83" aria-hidden="true" tabindex="-1"></a><span class="co"># get allele counts, was determined in all cells ----</span></span>
<span id="cb13-84"><a href="#cb13-84" aria-hidden="true" tabindex="-1"></a>dir<span class="ot">=</span><span class="st">"/scratch/kou/composite/numbat/"</span></span>
<span id="cb13-85"><a href="#cb13-85" aria-hidden="true" tabindex="-1"></a>files <span class="ot">=</span> <span class="fu">list.files</span>(<span class="fu">paste0</span>(dir,LN_ID), <span class="at">pattern =</span> <span class="st">"tsv.gz$"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb13-86"><a href="#cb13-86" aria-hidden="true" tabindex="-1"></a>alleleCounts <span class="ot">=</span> <span class="fu">lapply</span>(files, read_tsv)</span>
<span id="cb13-87"><a href="#cb13-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-88"><a href="#cb13-88" aria-hidden="true" tabindex="-1"></a><span class="do">## modify cell id to match seurat, filter for "keep" cells</span></span>
<span id="cb13-89"><a href="#cb13-89" aria-hidden="true" tabindex="-1"></a>alleleCounts <span class="ot">=</span> <span class="fu">map2</span>(</span>
<span id="cb13-90"><a href="#cb13-90" aria-hidden="true" tabindex="-1"></a>  LN_ID,</span>
<span id="cb13-91"><a href="#cb13-91" aria-hidden="true" tabindex="-1"></a>  alleleCounts, </span>
<span id="cb13-92"><a href="#cb13-92" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> .y <span class="sc">%&gt;%</span></span>
<span id="cb13-93"><a href="#cb13-93" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">cell =</span> <span class="fu">paste0</span>(.x,<span class="st">"_"</span>,cell)) <span class="sc">%&gt;%</span></span>
<span id="cb13-94"><a href="#cb13-94" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(cell <span class="sc">%in%</span> keep)</span>
<span id="cb13-95"><a href="#cb13-95" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-96"><a href="#cb13-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-97"><a href="#cb13-97" aria-hidden="true" tabindex="-1"></a><span class="co"># get count matrix ----</span></span>
<span id="cb13-98"><a href="#cb13-98" aria-hidden="true" tabindex="-1"></a><span class="do">## count matrix, split by sample</span></span>
<span id="cb13-99"><a href="#cb13-99" aria-hidden="true" tabindex="-1"></a>countMat <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts</span>
<span id="cb13-100"><a href="#cb13-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-101"><a href="#cb13-101" aria-hidden="true" tabindex="-1"></a><span class="do">## remove mitotic and singletons</span></span>
<span id="cb13-102"><a href="#cb13-102" aria-hidden="true" tabindex="-1"></a>countMat <span class="ot">=</span> countMat[,<span class="fu">colnames</span>(countMat) <span class="sc">%in%</span> keep]</span>
<span id="cb13-103"><a href="#cb13-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-104"><a href="#cb13-104" aria-hidden="true" tabindex="-1"></a><span class="do">## split count matrix by sample</span></span>
<span id="cb13-105"><a href="#cb13-105" aria-hidden="true" tabindex="-1"></a>countMat <span class="ot">=</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(x){</span>
<span id="cb13-106"><a href="#cb13-106" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> countMat[,<span class="fu">grepl</span>(LN_ID[[x]],<span class="fu">colnames</span>(countMat))]</span>
<span id="cb13-107"><a href="#cb13-107" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-108"><a href="#cb13-108" aria-hidden="true" tabindex="-1"></a>  <span class="co"># round mat to integer, otherwise error later</span></span>
<span id="cb13-109"><a href="#cb13-109" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> <span class="fu">round</span>(mat)</span>
<span id="cb13-110"><a href="#cb13-110" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-111"><a href="#cb13-111" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mat)</span>
<span id="cb13-112"><a href="#cb13-112" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb13-113"><a href="#cb13-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-114"><a href="#cb13-114" aria-hidden="true" tabindex="-1"></a><span class="co"># make collective reference matrix ----</span></span>
<span id="cb13-115"><a href="#cb13-115" aria-hidden="true" tabindex="-1"></a><span class="do">## remove bcl2.hi (likely) tumor cells</span></span>
<span id="cb13-116"><a href="#cb13-116" aria-hidden="true" tabindex="-1"></a>noBCL2hi <span class="ot">=</span> <span class="fu">levels</span>(wnn)[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"BCL2.hi|singleton|mitotic"</span>,<span class="fu">levels</span>(wnn))]</span>
<span id="cb13-117"><a href="#cb13-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-118"><a href="#cb13-118" aria-hidden="true" tabindex="-1"></a><span class="do">## get counts, subset cells </span></span>
<span id="cb13-119"><a href="#cb13-119" aria-hidden="true" tabindex="-1"></a>refCounts <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[, <span class="fu">Idents</span>(wnn) <span class="sc">%in%</span> noBCL2hi]</span>
<span id="cb13-120"><a href="#cb13-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-121"><a href="#cb13-121" aria-hidden="true" tabindex="-1"></a><span class="do">## create anno</span></span>
<span id="cb13-122"><a href="#cb13-122" aria-hidden="true" tabindex="-1"></a>refAnno <span class="ot">=</span> meta <span class="sc">%&gt;%</span> </span>
<span id="cb13-123"><a href="#cb13-123" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cell =</span> <span class="fu">paste0</span>(orig.ident,<span class="st">"_"</span>,cell_ids)) <span class="sc">%&gt;%</span></span>
<span id="cb13-124"><a href="#cb13-124" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cell <span class="sc">%in%</span> <span class="fu">colnames</span>(refCounts)) <span class="sc">%&gt;%</span></span>
<span id="cb13-125"><a href="#cb13-125" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(cell, <span class="at">group =</span> prelim_anno)</span>
<span id="cb13-126"><a href="#cb13-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-127"><a href="#cb13-127" aria-hidden="true" tabindex="-1"></a><span class="do">## aggregate counts</span></span>
<span id="cb13-128"><a href="#cb13-128" aria-hidden="true" tabindex="-1"></a>ref_internal <span class="ot">=</span> <span class="fu">aggregate_counts</span>(</span>
<span id="cb13-129"><a href="#cb13-129" aria-hidden="true" tabindex="-1"></a>  <span class="at">count_mat =</span> refCounts, </span>
<span id="cb13-130"><a href="#cb13-130" aria-hidden="true" tabindex="-1"></a>  <span class="at">annot =</span> refAnno, </span>
<span id="cb13-131"><a href="#cb13-131" aria-hidden="true" tabindex="-1"></a>  <span class="at">normalized =</span> T, </span>
<span id="cb13-132"><a href="#cb13-132" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> F</span>
<span id="cb13-133"><a href="#cb13-133" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-134"><a href="#cb13-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-135"><a href="#cb13-135" aria-hidden="true" tabindex="-1"></a><span class="co"># run numbat ----</span></span>
<span id="cb13-136"><a href="#cb13-136" aria-hidden="true" tabindex="-1"></a><span class="do">## should run on slurm next time</span></span>
<span id="cb13-137"><a href="#cb13-137" aria-hidden="true" tabindex="-1"></a>outdir<span class="ot">=</span><span class="st">"rdata/numbat/benchmark1/"</span></span>
<span id="cb13-138"><a href="#cb13-138" aria-hidden="true" tabindex="-1"></a>nbResults <span class="ot">=</span> <span class="fu">map</span>(</span>
<span id="cb13-139"><a href="#cb13-139" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(countMat),</span>
<span id="cb13-140"><a href="#cb13-140" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">run_numbat</span>(</span>
<span id="cb13-141"><a href="#cb13-141" aria-hidden="true" tabindex="-1"></a>    <span class="at">count_mat =</span> countMat[[.x]],</span>
<span id="cb13-142"><a href="#cb13-142" aria-hidden="true" tabindex="-1"></a>    <span class="at">lambdas_ref =</span> ref_internal,</span>
<span id="cb13-143"><a href="#cb13-143" aria-hidden="true" tabindex="-1"></a>    <span class="at">df_allele =</span> alleleCounts[[.x]],</span>
<span id="cb13-144"><a href="#cb13-144" aria-hidden="true" tabindex="-1"></a>    <span class="at">genome =</span> <span class="st">"hg38"</span>,</span>
<span id="cb13-145"><a href="#cb13-145" aria-hidden="true" tabindex="-1"></a>    <span class="at">out_dir =</span> <span class="fu">paste0</span>(outdir,LN_ID[[.x]]),</span>
<span id="cb13-146"><a href="#cb13-146" aria-hidden="true" tabindex="-1"></a>    <span class="at">t =</span> <span class="fl">1e-06</span>, </span>
<span id="cb13-147"><a href="#cb13-147" aria-hidden="true" tabindex="-1"></a>    <span class="at">gamma =</span> <span class="dv">20</span>,</span>
<span id="cb13-148"><a href="#cb13-148" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_cells =</span> <span class="dv">30</span>, </span>
<span id="cb13-149"><a href="#cb13-149" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_depth =</span> <span class="dv">0</span>,</span>
<span id="cb13-150"><a href="#cb13-150" aria-hidden="true" tabindex="-1"></a>    <span class="at">multi_allelic =</span> T,</span>
<span id="cb13-151"><a href="#cb13-151" aria-hidden="true" tabindex="-1"></a>    <span class="at">tau =</span> <span class="fl">0.25</span>,</span>
<span id="cb13-152"><a href="#cb13-152" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>,</span>
<span id="cb13-153"><a href="#cb13-153" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores_nni =</span> <span class="dv">1</span></span>
<span id="cb13-154"><a href="#cb13-154" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-155"><a href="#cb13-155" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-156"><a href="#cb13-156" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-157"><a href="#cb13-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-158"><a href="#cb13-158" aria-hidden="true" tabindex="-1"></a><span class="fu">### Benchmark 2</span></span>
<span id="cb13-159"><a href="#cb13-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-160"><a href="#cb13-160" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Allele / Count input: all cells (including singleton and mitotic!), separated by patient</span>
<span id="cb13-161"><a href="#cb13-161" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Expression reference: all cells minus singleton, mitotic, BCL2.hi B cells; used patient specific reference</span>
<span id="cb13-162"><a href="#cb13-162" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`t`</span> transition probabiltiy: 1e-6</span>
<span id="cb13-163"><a href="#cb13-163" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`tau`</span> = 0.25</span>
<span id="cb13-164"><a href="#cb13-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-167"><a href="#cb13-167" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-168"><a href="#cb13-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: benchmark2_code</span></span>
<span id="cb13-169"><a href="#cb13-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-170"><a href="#cb13-170" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- get allele counts, was determined in all cells ----</span></span>
<span id="cb13-171"><a href="#cb13-171" aria-hidden="true" tabindex="-1"></a>dir<span class="ot">=</span><span class="st">"/scratch/kou/composite/numbat/"</span></span>
<span id="cb13-172"><a href="#cb13-172" aria-hidden="true" tabindex="-1"></a>files <span class="ot">=</span> <span class="fu">list.files</span>(<span class="fu">paste0</span>(dir,LN_ID), <span class="at">pattern =</span> <span class="st">"tsv.gz$"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb13-173"><a href="#cb13-173" aria-hidden="true" tabindex="-1"></a>alleleCounts2 <span class="ot">=</span> <span class="fu">lapply</span>(files, read_tsv)</span>
<span id="cb13-174"><a href="#cb13-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-175"><a href="#cb13-175" aria-hidden="true" tabindex="-1"></a><span class="do">## modify cell id to match seurat</span></span>
<span id="cb13-176"><a href="#cb13-176" aria-hidden="true" tabindex="-1"></a>alleleCounts2 <span class="ot">=</span> <span class="fu">map2</span>(</span>
<span id="cb13-177"><a href="#cb13-177" aria-hidden="true" tabindex="-1"></a>  LN_ID,</span>
<span id="cb13-178"><a href="#cb13-178" aria-hidden="true" tabindex="-1"></a>  alleleCounts2, </span>
<span id="cb13-179"><a href="#cb13-179" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> .y <span class="sc">%&gt;%</span></span>
<span id="cb13-180"><a href="#cb13-180" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">cell =</span> <span class="fu">paste0</span>(.x,<span class="st">"_"</span>,cell))</span>
<span id="cb13-181"><a href="#cb13-181" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-182"><a href="#cb13-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-183"><a href="#cb13-183" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- get count matrix ----</span></span>
<span id="cb13-184"><a href="#cb13-184" aria-hidden="true" tabindex="-1"></a><span class="do">## count matrix, split by sample</span></span>
<span id="cb13-185"><a href="#cb13-185" aria-hidden="true" tabindex="-1"></a>countMat2 <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts</span>
<span id="cb13-186"><a href="#cb13-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-187"><a href="#cb13-187" aria-hidden="true" tabindex="-1"></a><span class="do">## split count matrix by sample</span></span>
<span id="cb13-188"><a href="#cb13-188" aria-hidden="true" tabindex="-1"></a>countMat2 <span class="ot">=</span> <span class="fu">future_lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(x){</span>
<span id="cb13-189"><a href="#cb13-189" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> countMat2[,<span class="fu">grepl</span>(LN_ID[[x]],<span class="fu">colnames</span>(countMat2))]</span>
<span id="cb13-190"><a href="#cb13-190" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-191"><a href="#cb13-191" aria-hidden="true" tabindex="-1"></a>  <span class="co"># round mat to integer, otherwise error later</span></span>
<span id="cb13-192"><a href="#cb13-192" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> <span class="fu">round</span>(mat)</span>
<span id="cb13-193"><a href="#cb13-193" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mat)</span>
<span id="cb13-194"><a href="#cb13-194" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb13-195"><a href="#cb13-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-196"><a href="#cb13-196" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- make patient specific reference matrix ----</span></span>
<span id="cb13-197"><a href="#cb13-197" aria-hidden="true" tabindex="-1"></a><span class="do">## remove bcl2.hi and mitotic bc (likely) tumor cells</span></span>
<span id="cb13-198"><a href="#cb13-198" aria-hidden="true" tabindex="-1"></a>noBCL2hi <span class="ot">=</span> <span class="fu">levels</span>(wnn)[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"BCL2.hi|mitotic"</span>,<span class="fu">levels</span>(wnn))]</span>
<span id="cb13-199"><a href="#cb13-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-200"><a href="#cb13-200" aria-hidden="true" tabindex="-1"></a><span class="do">## get counts, subset cells </span></span>
<span id="cb13-201"><a href="#cb13-201" aria-hidden="true" tabindex="-1"></a>refCounts2 <span class="ot">=</span> wnn[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[, <span class="fu">Idents</span>(wnn) <span class="sc">%in%</span> noBCL2hi]</span>
<span id="cb13-202"><a href="#cb13-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-203"><a href="#cb13-203" aria-hidden="true" tabindex="-1"></a><span class="do">## split ref matrix by sample</span></span>
<span id="cb13-204"><a href="#cb13-204" aria-hidden="true" tabindex="-1"></a>refCounts2 <span class="ot">=</span> <span class="fu">future_lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(x){</span>
<span id="cb13-205"><a href="#cb13-205" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">=</span> refCounts2[,<span class="fu">grepl</span>(LN_ID[[x]],<span class="fu">colnames</span>(refCounts2))]</span>
<span id="cb13-206"><a href="#cb13-206" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mat)</span>
<span id="cb13-207"><a href="#cb13-207" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb13-208"><a href="#cb13-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-209"><a href="#cb13-209" aria-hidden="true" tabindex="-1"></a><span class="do">## create anno </span></span>
<span id="cb13-210"><a href="#cb13-210" aria-hidden="true" tabindex="-1"></a>refAnno2 <span class="ot">=</span> meta <span class="sc">%&gt;%</span></span>
<span id="cb13-211"><a href="#cb13-211" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(prelim_anno <span class="sc">%in%</span> noBCL2hi) <span class="sc">%&gt;%</span></span>
<span id="cb13-212"><a href="#cb13-212" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"cell"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-213"><a href="#cb13-213" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(orig.ident) <span class="sc">%&gt;%</span></span>
<span id="cb13-214"><a href="#cb13-214" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(orig.ident, cell, <span class="at">group =</span> prelim_anno) <span class="sc">%&gt;%</span></span>
<span id="cb13-215"><a href="#cb13-215" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_split</span>(<span class="at">.keep =</span> F)</span>
<span id="cb13-216"><a href="#cb13-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-217"><a href="#cb13-217" aria-hidden="true" tabindex="-1"></a><span class="do">## aggregate counts</span></span>
<span id="cb13-218"><a href="#cb13-218" aria-hidden="true" tabindex="-1"></a>ref2 <span class="ot">=</span> <span class="fu">future_map2</span>(</span>
<span id="cb13-219"><a href="#cb13-219" aria-hidden="true" tabindex="-1"></a>  refCounts2, </span>
<span id="cb13-220"><a href="#cb13-220" aria-hidden="true" tabindex="-1"></a>  refAnno2,</span>
<span id="cb13-221"><a href="#cb13-221" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">aggregate_counts</span>(</span>
<span id="cb13-222"><a href="#cb13-222" aria-hidden="true" tabindex="-1"></a>    <span class="at">count_mat =</span> .x,</span>
<span id="cb13-223"><a href="#cb13-223" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot =</span> .y, </span>
<span id="cb13-224"><a href="#cb13-224" aria-hidden="true" tabindex="-1"></a>    <span class="at">normalized =</span> T, </span>
<span id="cb13-225"><a href="#cb13-225" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> F</span>
<span id="cb13-226"><a href="#cb13-226" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-227"><a href="#cb13-227" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-228"><a href="#cb13-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-229"><a href="#cb13-229" aria-hidden="true" tabindex="-1"></a><span class="co"># ---- run numbat ----</span></span>
<span id="cb13-230"><a href="#cb13-230" aria-hidden="true" tabindex="-1"></a>outdir<span class="ot">=</span><span class="st">"rdata/numbat/benchmark2/"</span></span>
<span id="cb13-231"><a href="#cb13-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-232"><a href="#cb13-232" aria-hidden="true" tabindex="-1"></a>nbResults2 <span class="ot">=</span> <span class="fu">future_map</span>(</span>
<span id="cb13-233"><a href="#cb13-233" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(countMat2),</span>
<span id="cb13-234"><a href="#cb13-234" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">run_numbat</span>(</span>
<span id="cb13-235"><a href="#cb13-235" aria-hidden="true" tabindex="-1"></a>    <span class="at">count_mat =</span> countMat2[[.x]],</span>
<span id="cb13-236"><a href="#cb13-236" aria-hidden="true" tabindex="-1"></a>    <span class="at">lambdas_ref =</span> ref2[[.x]],</span>
<span id="cb13-237"><a href="#cb13-237" aria-hidden="true" tabindex="-1"></a>    <span class="at">df_allele =</span> alleleCounts2[[.x]],</span>
<span id="cb13-238"><a href="#cb13-238" aria-hidden="true" tabindex="-1"></a>    <span class="at">genome =</span> <span class="st">"hg38"</span>,</span>
<span id="cb13-239"><a href="#cb13-239" aria-hidden="true" tabindex="-1"></a>    <span class="at">out_dir =</span> <span class="fu">paste0</span>(outdir,LN_ID[[.x]]),</span>
<span id="cb13-240"><a href="#cb13-240" aria-hidden="true" tabindex="-1"></a>    <span class="at">t =</span> <span class="fl">1e-06</span>, </span>
<span id="cb13-241"><a href="#cb13-241" aria-hidden="true" tabindex="-1"></a>    <span class="at">gamma =</span> <span class="dv">20</span>,</span>
<span id="cb13-242"><a href="#cb13-242" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_cells =</span> <span class="dv">30</span>, </span>
<span id="cb13-243"><a href="#cb13-243" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_depth =</span> <span class="dv">0</span>,</span>
<span id="cb13-244"><a href="#cb13-244" aria-hidden="true" tabindex="-1"></a>    <span class="at">multi_allelic =</span> T,</span>
<span id="cb13-245"><a href="#cb13-245" aria-hidden="true" tabindex="-1"></a>    <span class="at">tau =</span> <span class="fl">0.25</span>,</span>
<span id="cb13-246"><a href="#cb13-246" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">4</span>,</span>
<span id="cb13-247"><a href="#cb13-247" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores_nni =</span> <span class="dv">4</span></span>
<span id="cb13-248"><a href="#cb13-248" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-249"><a href="#cb13-249" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-250"><a href="#cb13-250" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-251"><a href="#cb13-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-252"><a href="#cb13-252" aria-hidden="true" tabindex="-1"></a><span class="fu">### Benchmark 3</span></span>
<span id="cb13-253"><a href="#cb13-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-254"><a href="#cb13-254" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>reference from rLN? needs to be same anno?</span>
<span id="cb13-255"><a href="#cb13-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-256"><a href="#cb13-256" aria-hidden="true" tabindex="-1"></a><span class="fu">## Results</span></span>
<span id="cb13-257"><a href="#cb13-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-258"><a href="#cb13-258" aria-hidden="true" tabindex="-1"></a>**Import numbat results**</span>
<span id="cb13-259"><a href="#cb13-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-262"><a href="#cb13-262" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-263"><a href="#cb13-263" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: result_import</span></span>
<span id="cb13-264"><a href="#cb13-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-265"><a href="#cb13-265" aria-hidden="true" tabindex="-1"></a><span class="co"># initialize numbat class</span></span>
<span id="cb13-266"><a href="#cb13-266" aria-hidden="true" tabindex="-1"></a>dir1 <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"rdata/numbat/benchmark1/"</span>,LN_ID)</span>
<span id="cb13-267"><a href="#cb13-267" aria-hidden="true" tabindex="-1"></a>nbResults1 <span class="ot">=</span> <span class="fu">lapply</span>(dir1, Numbat<span class="sc">$</span>new)</span>
<span id="cb13-268"><a href="#cb13-268" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(nbResults1) <span class="ot">=</span> LN_ID</span>
<span id="cb13-269"><a href="#cb13-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-270"><a href="#cb13-270" aria-hidden="true" tabindex="-1"></a>dir2 <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"rdata/numbat/benchmark2/"</span>,LN_ID)</span>
<span id="cb13-271"><a href="#cb13-271" aria-hidden="true" tabindex="-1"></a>nbResults2 <span class="ot">=</span> <span class="fu">lapply</span>(dir2, Numbat<span class="sc">$</span>new)</span>
<span id="cb13-272"><a href="#cb13-272" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(nbResults2) <span class="ot">=</span> LN_ID</span>
<span id="cb13-273"><a href="#cb13-273" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-274"><a href="#cb13-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-275"><a href="#cb13-275" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compare pseudobulked clones (output of `run_numbat`)</span></span>
<span id="cb13-276"><a href="#cb13-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-279"><a href="#cb13-279" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-280"><a href="#cb13-280" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: pseudobulked_clones</span></span>
<span id="cb13-281"><a href="#cb13-281" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 1</span></span>
<span id="cb13-282"><a href="#cb13-282" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb13-283"><a href="#cb13-283" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-subcap:</span></span>
<span id="cb13-284"><a href="#cb13-284" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0025, bench. 1"</span></span>
<span id="cb13-285"><a href="#cb13-285" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0025, bench. 2"</span></span>
<span id="cb13-286"><a href="#cb13-286" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0177, bench. 1"</span></span>
<span id="cb13-287"><a href="#cb13-287" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0177, bench. 2"</span></span>
<span id="cb13-288"><a href="#cb13-288" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0193, bench. 1"</span></span>
<span id="cb13-289"><a href="#cb13-289" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0193, bench. 2"</span></span>
<span id="cb13-290"><a href="#cb13-290" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0438, bench. 1"</span></span>
<span id="cb13-291"><a href="#cb13-291" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "LN0438, bench. 2"</span></span>
<span id="cb13-292"><a href="#cb13-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-293"><a href="#cb13-293" aria-hidden="true" tabindex="-1"></a><span class="co"># need absolute paths!</span></span>
<span id="cb13-294"><a href="#cb13-294" aria-hidden="true" tabindex="-1"></a>LN_ID <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LN0025"</span>,<span class="st">"LN0177"</span>,<span class="st">"LN0193"</span>,<span class="st">"LN0438"</span>)</span>
<span id="cb13-295"><a href="#cb13-295" aria-hidden="true" tabindex="-1"></a>plotpaths <span class="ot">=</span> purrr<span class="sc">::</span><span class="fu">map</span>(</span>
<span id="cb13-296"><a href="#cb13-296" aria-hidden="true" tabindex="-1"></a>  LN_ID, </span>
<span id="cb13-297"><a href="#cb13-297" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">file.path</span>(</span>
<span id="cb13-298"><a href="#cb13-298" aria-hidden="true" tabindex="-1"></a>    here<span class="sc">::</span><span class="fu">here</span>(),</span>
<span id="cb13-299"><a href="#cb13-299" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"rdata/numbat/benchmark"</span>, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)),</span>
<span id="cb13-300"><a href="#cb13-300" aria-hidden="true" tabindex="-1"></a>    .x,</span>
<span id="cb13-301"><a href="#cb13-301" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bulk_clones_final.png"</span></span>
<span id="cb13-302"><a href="#cb13-302" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-303"><a href="#cb13-303" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-304"><a href="#cb13-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-305"><a href="#cb13-305" aria-hidden="true" tabindex="-1"></a><span class="co"># lapply didn't work here</span></span>
<span id="cb13-306"><a href="#cb13-306" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">1</span>]])</span>
<span id="cb13-307"><a href="#cb13-307" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">2</span>]])</span>
<span id="cb13-308"><a href="#cb13-308" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">3</span>]])</span>
<span id="cb13-309"><a href="#cb13-309" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(plotpaths[[<span class="dv">4</span>]])</span>
<span id="cb13-310"><a href="#cb13-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-311"><a href="#cb13-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-312"><a href="#cb13-312" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>With the exception of LN0025, there were no major differences between benchmark 1 and 2.</span>
<span id="cb13-313"><a href="#cb13-313" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Continued the rest of the analysis with benchmark 2 because more cells were included in analysis (ie - mitotic and singletons).</span>
<span id="cb13-314"><a href="#cb13-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-315"><a href="#cb13-315" aria-hidden="true" tabindex="-1"></a><span class="fu">### Single cell posterior probabilities for each CNV</span></span>
<span id="cb13-316"><a href="#cb13-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-317"><a href="#cb13-317" aria-hidden="true" tabindex="-1"></a><span class="sc">\[\[</span>TO DO: add more detail + plots, or remove<span class="sc">\]\]</span></span>
<span id="cb13-318"><a href="#cb13-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-319"><a href="#cb13-319" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>p_cnv: etc</span>
<span id="cb13-320"><a href="#cb13-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-323"><a href="#cb13-323" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-324"><a href="#cb13-324" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: cnv_post_prob</span></span>
<span id="cb13-325"><a href="#cb13-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-326"><a href="#cb13-326" aria-hidden="true" tabindex="-1"></a><span class="co"># get current joint_post</span></span>
<span id="cb13-327"><a href="#cb13-327" aria-hidden="true" tabindex="-1"></a>jointPost <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb13-328"><a href="#cb13-328" aria-hidden="true" tabindex="-1"></a>  nbResults2, </span>
<span id="cb13-329"><a href="#cb13-329" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb) {</span>
<span id="cb13-330"><a href="#cb13-330" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-331"><a href="#cb13-331" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get probabilities for each cnv</span></span>
<span id="cb13-332"><a href="#cb13-332" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> <span class="fu">as_tibble</span>(nb<span class="sc">$</span>joint_post) <span class="sc">%&gt;%</span></span>
<span id="cb13-333"><a href="#cb13-333" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">cnv =</span> <span class="fu">factor</span>(<span class="fu">paste0</span>(seg,<span class="st">"_"</span>,cnv_state))) <span class="sc">%&gt;%</span></span>
<span id="cb13-334"><a href="#cb13-334" aria-hidden="true" tabindex="-1"></a>      <span class="co"># select(</span></span>
<span id="cb13-335"><a href="#cb13-335" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   cell,cnv,cnv_state,LLR,</span></span>
<span id="cb13-336"><a href="#cb13-336" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   CHROM,seg_start,seg_end,</span></span>
<span id="cb13-337"><a href="#cb13-337" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   n_genes,n_snp,</span></span>
<span id="cb13-338"><a href="#cb13-338" aria-hidden="true" tabindex="-1"></a>      <span class="co">#   p_cnv,p_cnv_x,p_cnv_y</span></span>
<span id="cb13-339"><a href="#cb13-339" aria-hidden="true" tabindex="-1"></a>      <span class="co"># )</span></span>
<span id="cb13-340"><a href="#cb13-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-341"><a href="#cb13-341" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb13-342"><a href="#cb13-342" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb13-343"><a href="#cb13-343" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-344"><a href="#cb13-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-345"><a href="#cb13-345" aria-hidden="true" tabindex="-1"></a><span class="co"># add cell annotation</span></span>
<span id="cb13-346"><a href="#cb13-346" aria-hidden="true" tabindex="-1"></a>jointPost <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb13-347"><a href="#cb13-347" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(jointPost), </span>
<span id="cb13-348"><a href="#cb13-348" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) {</span>
<span id="cb13-349"><a href="#cb13-349" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get cell anno</span></span>
<span id="cb13-350"><a href="#cb13-350" aria-hidden="true" tabindex="-1"></a>    cellanno <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb13-351"><a href="#cb13-351" aria-hidden="true" tabindex="-1"></a>      <span class="at">cell =</span> <span class="fu">rownames</span>(meta),</span>
<span id="cb13-352"><a href="#cb13-352" aria-hidden="true" tabindex="-1"></a>      <span class="at">major_type =</span> meta<span class="sc">$</span>major_type,</span>
<span id="cb13-353"><a href="#cb13-353" aria-hidden="true" tabindex="-1"></a>      <span class="at">prelim_anno =</span> meta<span class="sc">$</span>prelim_anno,</span>
<span id="cb13-354"><a href="#cb13-354" aria-hidden="true" tabindex="-1"></a>      <span class="at">tricycle =</span> meta<span class="sc">$</span>tricycle</span>
<span id="cb13-355"><a href="#cb13-355" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb13-356"><a href="#cb13-356" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-357"><a href="#cb13-357" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> <span class="fu">left_join</span>(jointPost[[x]], cellanno, <span class="at">by =</span> <span class="st">"cell"</span>)</span>
<span id="cb13-358"><a href="#cb13-358" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb13-359"><a href="#cb13-359" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb13-360"><a href="#cb13-360" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-361"><a href="#cb13-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-362"><a href="#cb13-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-363"><a href="#cb13-363" aria-hidden="true" tabindex="-1"></a><span class="co"># test plot, B cells only</span></span>
<span id="cb13-364"><a href="#cb13-364" aria-hidden="true" tabindex="-1"></a>jointPost[[<span class="dv">3</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb13-365"><a href="#cb13-365" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb13-366"><a href="#cb13-366" aria-hidden="true" tabindex="-1"></a>    <span class="co">#major_type %in% c("T"),</span></span>
<span id="cb13-367"><a href="#cb13-367" aria-hidden="true" tabindex="-1"></a>    cnv_state <span class="sc">!=</span> <span class="st">"loh"</span>,</span>
<span id="cb13-368"><a href="#cb13-368" aria-hidden="true" tabindex="-1"></a>    LLR <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb13-369"><a href="#cb13-369" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb13-370"><a href="#cb13-370" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">bcl2 =</span> <span class="fu">if_else</span>(<span class="fu">grepl</span>(<span class="st">"BCL2.hi"</span>, prelim_anno),<span class="st">"hi"</span>,<span class="st">"lo"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb13-371"><a href="#cb13-371" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(p_neu)) <span class="sc">+</span></span>
<span id="cb13-372"><a href="#cb13-372" aria-hidden="true" tabindex="-1"></a>  <span class="co">#geom_density() +</span></span>
<span id="cb13-373"><a href="#cb13-373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">color =</span> major_type)) <span class="sc">+</span></span>
<span id="cb13-374"><a href="#cb13-374" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(cnv), <span class="at">scales =</span> <span class="st">"free"</span>)</span>
<span id="cb13-375"><a href="#cb13-375" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-376"><a href="#cb13-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-377"><a href="#cb13-377" aria-hidden="true" tabindex="-1"></a><span class="fu">### Genotypes of clones</span></span>
<span id="cb13-378"><a href="#cb13-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-379"><a href="#cb13-379" aria-hidden="true" tabindex="-1"></a>TO DO:</span>
<span id="cb13-380"><a href="#cb13-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-381"><a href="#cb13-381" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>change levels to include type of mutation!!</span>
<span id="cb13-382"><a href="#cb13-382" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>re order levels, change color code!! resize ggplot</span>
<span id="cb13-383"><a href="#cb13-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-386"><a href="#cb13-386" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-387"><a href="#cb13-387" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: genotype_clones</span></span>
<span id="cb13-388"><a href="#cb13-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-389"><a href="#cb13-389" aria-hidden="true" tabindex="-1"></a><span class="co"># default results ----</span></span>
<span id="cb13-390"><a href="#cb13-390" aria-hidden="true" tabindex="-1"></a><span class="do">## get clone info</span></span>
<span id="cb13-391"><a href="#cb13-391" aria-hidden="true" tabindex="-1"></a>clonePost <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb13-392"><a href="#cb13-392" aria-hidden="true" tabindex="-1"></a>  nbResults2,</span>
<span id="cb13-393"><a href="#cb13-393" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb) {</span>
<span id="cb13-394"><a href="#cb13-394" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> nb<span class="sc">$</span>clone_post</span>
<span id="cb13-395"><a href="#cb13-395" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb13-396"><a href="#cb13-396" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb13-397"><a href="#cb13-397" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-398"><a href="#cb13-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-399"><a href="#cb13-399" aria-hidden="true" tabindex="-1"></a>clones <span class="ot">=</span> <span class="fu">do.call</span>(rbind,clonePost)</span>
<span id="cb13-400"><a href="#cb13-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-401"><a href="#cb13-401" aria-hidden="true" tabindex="-1"></a><span class="do">## checked order first all(clones$cell == rownames(embeds))</span></span>
<span id="cb13-402"><a href="#cb13-402" aria-hidden="true" tabindex="-1"></a>clones <span class="ot">=</span> <span class="fu">cbind</span>(clones, embeds)</span>
<span id="cb13-403"><a href="#cb13-403" aria-hidden="true" tabindex="-1"></a>clones <span class="ot">=</span> <span class="fu">cbind</span>(clones, meta)</span>
<span id="cb13-404"><a href="#cb13-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-405"><a href="#cb13-405" aria-hidden="true" tabindex="-1"></a><span class="do">## relevel</span></span>
<span id="cb13-406"><a href="#cb13-406" aria-hidden="true" tabindex="-1"></a>clones<span class="sc">$</span>GT_opt <span class="ot">=</span> <span class="fu">factor</span>(</span>
<span id="cb13-407"><a href="#cb13-407" aria-hidden="true" tabindex="-1"></a>  clones<span class="sc">$</span>GT_opt, </span>
<span id="cb13-408"><a href="#cb13-408" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">""</span>,<span class="st">""</span>)</span>
<span id="cb13-409"><a href="#cb13-409" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-410"><a href="#cb13-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-411"><a href="#cb13-411" aria-hidden="true" tabindex="-1"></a><span class="do">## genotype color code</span></span>
<span id="cb13-412"><a href="#cb13-412" aria-hidden="true" tabindex="-1"></a>gtColor <span class="ot">=</span> <span class="fu">hcl.colors</span>(<span class="at">n =</span> <span class="dv">4</span>, <span class="st">"Spectral"</span>)</span>
<span id="cb13-413"><a href="#cb13-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-414"><a href="#cb13-414" aria-hidden="true" tabindex="-1"></a><span class="do">## generate random order for plotting</span></span>
<span id="cb13-415"><a href="#cb13-415" aria-hidden="true" tabindex="-1"></a>cloneplot1 <span class="ot">=</span> clones[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(clones), <span class="fu">nrow</span>(clones), <span class="at">replace =</span> F),] <span class="sc">%&gt;%</span></span>
<span id="cb13-416"><a href="#cb13-416" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(wnnUMAP_1, wnnUMAP_2))<span class="sc">+</span></span>
<span id="cb13-417"><a href="#cb13-417" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(</span>
<span id="cb13-418"><a href="#cb13-418" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">color =</span> GT_opt), </span>
<span id="cb13-419"><a href="#cb13-419" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.05</span>, </span>
<span id="cb13-420"><a href="#cb13-420" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb13-421"><a href="#cb13-421" aria-hidden="true" tabindex="-1"></a>  )<span class="sc">+</span></span>
<span id="cb13-422"><a href="#cb13-422" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_umap</span>() <span class="sc">+</span></span>
<span id="cb13-423"><a href="#cb13-423" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#F5F5F5"</span>,gtColor))<span class="sc">+</span></span>
<span id="cb13-424"><a href="#cb13-424" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"Genotype"</span>) <span class="sc">+</span></span>
<span id="cb13-425"><a href="#cb13-425" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(orig.ident))</span>
<span id="cb13-426"><a href="#cb13-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-427"><a href="#cb13-427" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(</span>
<span id="cb13-428"><a href="#cb13-428" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">"figures/"</span>,<span class="fu">format</span>(<span class="fu">Sys.time</span>(),<span class="st">"%y%m%d"</span>),<span class="st">"_NumbatClonesK1.png"</span>),</span>
<span id="cb13-429"><a href="#cb13-429" aria-hidden="true" tabindex="-1"></a>  cloneplot1, <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">3</span></span>
<span id="cb13-430"><a href="#cb13-430" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-431"><a href="#cb13-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-432"><a href="#cb13-432" aria-hidden="true" tabindex="-1"></a><span class="co"># recut tree (k = 2) to see if subclones resolve ----</span></span>
<span id="cb13-433"><a href="#cb13-433" aria-hidden="true" tabindex="-1"></a><span class="do">## this will change the numbat benchmark2 dir</span></span>
<span id="cb13-434"><a href="#cb13-434" aria-hidden="true" tabindex="-1"></a>nb_k2 <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb13-435"><a href="#cb13-435" aria-hidden="true" tabindex="-1"></a>  nbResults2,</span>
<span id="cb13-436"><a href="#cb13-436" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb){</span>
<span id="cb13-437"><a href="#cb13-437" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-438"><a href="#cb13-438" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get 3 groups from 2 cuts</span></span>
<span id="cb13-439"><a href="#cb13-439" aria-hidden="true" tabindex="-1"></a>    nb<span class="sc">$</span><span class="fu">cutree</span>(<span class="at">n_cut =</span> <span class="dv">2</span>)</span>
<span id="cb13-440"><a href="#cb13-440" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(nb)</span>
<span id="cb13-441"><a href="#cb13-441" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb13-442"><a href="#cb13-442" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-443"><a href="#cb13-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-444"><a href="#cb13-444" aria-hidden="true" tabindex="-1"></a><span class="do">## get clone info</span></span>
<span id="cb13-445"><a href="#cb13-445" aria-hidden="true" tabindex="-1"></a>clonePost_k2 <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb13-446"><a href="#cb13-446" aria-hidden="true" tabindex="-1"></a>  nb_k2,</span>
<span id="cb13-447"><a href="#cb13-447" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(nb) {</span>
<span id="cb13-448"><a href="#cb13-448" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> nb<span class="sc">$</span>clone_post</span>
<span id="cb13-449"><a href="#cb13-449" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb13-450"><a href="#cb13-450" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb13-451"><a href="#cb13-451" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-452"><a href="#cb13-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-453"><a href="#cb13-453" aria-hidden="true" tabindex="-1"></a>clones_k2 <span class="ot">=</span> <span class="fu">do.call</span>(rbind,clonePost_k2)</span>
<span id="cb13-454"><a href="#cb13-454" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-455"><a href="#cb13-455" aria-hidden="true" tabindex="-1"></a><span class="do">## checked order first all(clones$cell == rownames(embeds))</span></span>
<span id="cb13-456"><a href="#cb13-456" aria-hidden="true" tabindex="-1"></a>clones_k2 <span class="ot">=</span> <span class="fu">cbind</span>(clones_k2, embeds)</span>
<span id="cb13-457"><a href="#cb13-457" aria-hidden="true" tabindex="-1"></a>clones_k2 <span class="ot">=</span> <span class="fu">cbind</span>(clones_k2,meta)</span>
<span id="cb13-458"><a href="#cb13-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-459"><a href="#cb13-459" aria-hidden="true" tabindex="-1"></a><span class="do">## genotype color code</span></span>
<span id="cb13-460"><a href="#cb13-460" aria-hidden="true" tabindex="-1"></a>gtColor <span class="ot">=</span> <span class="fu">hcl.colors</span>(<span class="at">n =</span> <span class="dv">8</span>, <span class="st">"Zissou"</span>)</span>
<span id="cb13-461"><a href="#cb13-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-462"><a href="#cb13-462" aria-hidden="true" tabindex="-1"></a><span class="do">## random order for plotting</span></span>
<span id="cb13-463"><a href="#cb13-463" aria-hidden="true" tabindex="-1"></a>cloneplot2 <span class="ot">=</span> clones_k2[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(clones_k2), <span class="fu">nrow</span>(clones_k2), <span class="at">replace =</span> F),] <span class="sc">%&gt;%</span></span>
<span id="cb13-464"><a href="#cb13-464" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(wnnUMAP_1, wnnUMAP_2))<span class="sc">+</span></span>
<span id="cb13-465"><a href="#cb13-465" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(</span>
<span id="cb13-466"><a href="#cb13-466" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">color =</span> GT_opt), </span>
<span id="cb13-467"><a href="#cb13-467" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.05</span>, </span>
<span id="cb13-468"><a href="#cb13-468" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb13-469"><a href="#cb13-469" aria-hidden="true" tabindex="-1"></a>  )<span class="sc">+</span></span>
<span id="cb13-470"><a href="#cb13-470" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_umap</span>() <span class="sc">+</span></span>
<span id="cb13-471"><a href="#cb13-471" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#F5F5F5"</span>,gtColor))<span class="sc">+</span></span>
<span id="cb13-472"><a href="#cb13-472" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"Genotype"</span>) <span class="sc">+</span></span>
<span id="cb13-473"><a href="#cb13-473" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(orig.ident))</span>
<span id="cb13-474"><a href="#cb13-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-475"><a href="#cb13-475" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(</span>
<span id="cb13-476"><a href="#cb13-476" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">"figures/"</span>,<span class="fu">format</span>(<span class="fu">Sys.time</span>(),<span class="st">"%y%m%d"</span>),<span class="st">"_NumbatClonesK2.png"</span>),</span>
<span id="cb13-477"><a href="#cb13-477" aria-hidden="true" tabindex="-1"></a>  cloneplot2, <span class="at">width =</span> <span class="dv">6</span>, <span class="at">height =</span> <span class="dv">3</span></span>
<span id="cb13-478"><a href="#cb13-478" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-479"><a href="#cb13-479" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-480"><a href="#cb13-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-483"><a href="#cb13-483" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-484"><a href="#cb13-484" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: genotypes_plot</span></span>
<span id="cb13-485"><a href="#cb13-485" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb13-486"><a href="#cb13-486" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 1</span></span>
<span id="cb13-487"><a href="#cb13-487" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-subcap:</span></span>
<span id="cb13-488"><a href="#cb13-488" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "Genotypes, ncut k: 1"</span></span>
<span id="cb13-489"><a href="#cb13-489" aria-hidden="true" tabindex="-1"></a><span class="co">#|  - "Genotypes, ncut k: 2"</span></span>
<span id="cb13-490"><a href="#cb13-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-491"><a href="#cb13-491" aria-hidden="true" tabindex="-1"></a>path <span class="ot">=</span> <span class="fu">list.files</span>(</span>
<span id="cb13-492"><a href="#cb13-492" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> here<span class="sc">::</span><span class="fu">here</span>(),</span>
<span id="cb13-493"><a href="#cb13-493" aria-hidden="true" tabindex="-1"></a>  <span class="at">pattern =</span> <span class="st">"NumbatClones"</span>,</span>
<span id="cb13-494"><a href="#cb13-494" aria-hidden="true" tabindex="-1"></a>  <span class="at">recursive =</span> T,</span>
<span id="cb13-495"><a href="#cb13-495" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.names =</span> T</span>
<span id="cb13-496"><a href="#cb13-496" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-497"><a href="#cb13-497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-498"><a href="#cb13-498" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(path[[<span class="dv">1</span>]])</span>
<span id="cb13-499"><a href="#cb13-499" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(path[[<span class="dv">2</span>]])</span>
<span id="cb13-500"><a href="#cb13-500" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-501"><a href="#cb13-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-502"><a href="#cb13-502" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>More background in T cells when forcing k = 2 (except possibly for LN0177)</span>
<span id="cb13-503"><a href="#cb13-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-504"><a href="#cb13-504" aria-hidden="true" tabindex="-1"></a>**Tumor vs non-tumor calls**</span>
<span id="cb13-505"><a href="#cb13-505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-506"><a href="#cb13-506" aria-hidden="true" tabindex="-1"></a>TO DO:</span>
<span id="cb13-507"><a href="#cb13-507" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-508"><a href="#cb13-508" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>update with updated GT with type of CNV</span>
<span id="cb13-509"><a href="#cb13-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-512"><a href="#cb13-512" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-513"><a href="#cb13-513" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: tumor_cell_calls</span></span>
<span id="cb13-514"><a href="#cb13-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-515"><a href="#cb13-515" aria-hidden="true" tabindex="-1"></a><span class="co"># checked same order w/ all()</span></span>
<span id="cb13-516"><a href="#cb13-516" aria-hidden="true" tabindex="-1"></a>tumorCalls <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb13-517"><a href="#cb13-517" aria-hidden="true" tabindex="-1"></a>  <span class="at">compartment_k1 =</span> clones<span class="sc">$</span>compartment_opt,</span>
<span id="cb13-518"><a href="#cb13-518" aria-hidden="true" tabindex="-1"></a>  <span class="at">gt_k1 =</span> clones<span class="sc">$</span>GT_opt,</span>
<span id="cb13-519"><a href="#cb13-519" aria-hidden="true" tabindex="-1"></a>  <span class="at">clone_k1 =</span> clones<span class="sc">$</span>clone_opt,</span>
<span id="cb13-520"><a href="#cb13-520" aria-hidden="true" tabindex="-1"></a>  <span class="at">compartment_k2 =</span> clones_k2<span class="sc">$</span>compartment_opt,</span>
<span id="cb13-521"><a href="#cb13-521" aria-hidden="true" tabindex="-1"></a>  <span class="at">gt_k2 =</span> clones_k2<span class="sc">$</span>GT_opt,</span>
<span id="cb13-522"><a href="#cb13-522" aria-hidden="true" tabindex="-1"></a>  <span class="at">clone_k2 =</span> clones_k2<span class="sc">$</span>clone_opt,</span>
<span id="cb13-523"><a href="#cb13-523" aria-hidden="true" tabindex="-1"></a>  <span class="at">major_type =</span> meta<span class="sc">$</span>major_type,</span>
<span id="cb13-524"><a href="#cb13-524" aria-hidden="true" tabindex="-1"></a>  <span class="at">prelim_anno =</span> meta<span class="sc">$</span>prelim_anno,</span>
<span id="cb13-525"><a href="#cb13-525" aria-hidden="true" tabindex="-1"></a>  <span class="at">orig.ident =</span> meta<span class="sc">$</span>orig.ident</span>
<span id="cb13-526"><a href="#cb13-526" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-527"><a href="#cb13-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-528"><a href="#cb13-528" aria-hidden="true" tabindex="-1"></a><span class="co"># pivot</span></span>
<span id="cb13-529"><a href="#cb13-529" aria-hidden="true" tabindex="-1"></a>tumorCalls <span class="ot">=</span> <span class="fu">pivot_longer</span>(</span>
<span id="cb13-530"><a href="#cb13-530" aria-hidden="true" tabindex="-1"></a>  tumorCalls,</span>
<span id="cb13-531"><a href="#cb13-531" aria-hidden="true" tabindex="-1"></a>  <span class="at">cols =</span> <span class="fu">names</span>(tumorCalls)[<span class="fu">grep</span>(<span class="st">"compartment"</span>,<span class="fu">names</span>(tumorCalls))],</span>
<span id="cb13-532"><a href="#cb13-532" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to =</span> <span class="st">"k"</span>,</span>
<span id="cb13-533"><a href="#cb13-533" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to =</span> <span class="st">"call"</span></span>
<span id="cb13-534"><a href="#cb13-534" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-535"><a href="#cb13-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-536"><a href="#cb13-536" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate proportion</span></span>
<span id="cb13-537"><a href="#cb13-537" aria-hidden="true" tabindex="-1"></a>propTumor <span class="ot">=</span> tumorCalls <span class="sc">%&gt;%</span></span>
<span id="cb13-538"><a href="#cb13-538" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(orig.ident, k, major_type, call) <span class="sc">%&gt;%</span></span>
<span id="cb13-539"><a href="#cb13-539" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb13-540"><a href="#cb13-540" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>(call) <span class="sc">%&gt;%</span></span>
<span id="cb13-541"><a href="#cb13-541" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prop =</span> n <span class="sc">/</span> <span class="fu">sum</span>(n))</span>
<span id="cb13-542"><a href="#cb13-542" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-543"><a href="#cb13-543" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-544"><a href="#cb13-544" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(propTumor, <span class="fu">aes</span>(k, prop))<span class="sc">+</span></span>
<span id="cb13-545"><a href="#cb13-545" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">fill =</span> call)) <span class="sc">+</span></span>
<span id="cb13-546"><a href="#cb13-546" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">rows =</span> <span class="fu">vars</span>(orig.ident),</span>
<span id="cb13-547"><a href="#cb13-547" aria-hidden="true" tabindex="-1"></a>             <span class="at">cols =</span> <span class="fu">vars</span>(major_type)) <span class="sc">+</span></span>
<span id="cb13-548"><a href="#cb13-548" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb13-549"><a href="#cb13-549" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-550"><a href="#cb13-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-551"><a href="#cb13-551" aria-hidden="true" tabindex="-1"></a>TO DO: more systematic way of taking all benchmark, compare n cut 1 vs 2, then assess clones</span>
<span id="cb13-552"><a href="#cb13-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-553"><a href="#cb13-553" aria-hidden="true" tabindex="-1"></a><span class="fu">## Annotate cells</span></span>
<span id="cb13-554"><a href="#cb13-554" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-555"><a href="#cb13-555" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>For now, annotate cells</span>
<span id="cb13-556"><a href="#cb13-556" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-559"><a href="#cb13-559" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-560"><a href="#cb13-560" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-561"><a href="#cb13-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-562"><a href="#cb13-562" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb13-563"><a href="#cb13-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-566"><a href="#cb13-566" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb13-567"><a href="#cb13-567" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: sessioninfo</span></span>
<span id="cb13-568"><a href="#cb13-568" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb13-569"><a href="#cb13-569" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb13-570"><a href="#cb13-570" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb13-571"><a href="#cb13-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-572"><a href="#cb13-572" aria-hidden="true" tabindex="-1"></a><span class="fu">## References</span></span>
<span id="cb13-573"><a href="#cb13-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-574"><a href="#cb13-574" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>original publication: <span class="ot">&lt;https://www.nature.com/articles/s41587-022-01468-y&gt;</span></span>
<span id="cb13-575"><a href="#cb13-575" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="ot">&lt;https://kharchenkolab.github.io/numbat/articles/numbat.html&gt;</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>